src/main/java/io/rocketpartners/cloud/utils/SqlUtils.java

```diff
@@ -431,6 +431,60 @@ public static String buildInsertSQL(Connection conn, String tableName, Object[]
       return sql.toString();
    }
 
+   public static String buildInsertOnDuplicateKeySQL(Connection conn, String tableName, Object[] columnNameArray)
+   {
+      StringBuffer sql = new StringBuffer(buildInsertSQL(conn, tableName, columnNameArray));
+      sql.append(" ON DUPLICATE KEY UPDATE ");
+      for (Object col : columnNameArray)
+      {
+         sql.append(col).append("=VALUES(").append(col).append("), ");
+      }
+      sql.substring(0, sql.length() - 1);
+      return sql.toString();
+   }
+
+   public static String mysqlUpsert(Connection conn, String tableName, List<Map<String, Object>> rows) throws Exception
+   {
+      LinkedHashSet keySet = new LinkedHashSet();
+
+      for (Map row : rows)
+      {
+         keySet.addAll(row.keySet());
+      }
+
+      List<String> keys = new ArrayList(keySet);
+      String sql = buildInsertOnDuplicateKeySQL(conn, tableName, keys.toArray());
+
+      Exception ex = null;
+      PreparedStatement stmt = conn.prepareStatement(sql);
+      try
+      {
+         notifyBefore("mysqlUpsert", sql, rows);
```

Comment:
I like that you saw/copied the notify pattern.  In this case, we should notify on "upsert" not on a specific db version of upsert.